src/data/list/to_finsupp.lean

```diff
@@ -0,0 +1,142 @@
+/-
+Copyright (c) 2022 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+
+import data.finsupp.basic
+
+/-!
+
+# Lists as finsupp
+
+# Main definitions
+
+- `list.to_finsupp`: Interpret a list as a finitely supported function, where the indexing type
+is `ℕ`, and the values are either the elements of the list (accessing by indexing) or `0` outside
+of the list.
+
+# Main theorems
+
+- `list.to_finsupp_eq_sum_map_enum_single`: A `l : list M` over `M` an `add_monoid`,
+when interpreted as a finitely supported function, is equal to the sum of `finsupp.single`
+produced by mapping over `list.enum l`.
+
+## Implementation details
+
+The functions defined here rely on a decidability predicate that each element in the list
+can be decidably determined to be not equal to zero or that one can decide one is out of the
+bounds of a list. For concretely defined lists that are made up of elements of decidable terms,
+this holds. More work will be needed to support lists over non-dec-eq types like `ℝ`, where the
+elements are beyond the dec-eq terms of casted values from `ℕ, ℤ, ℚ`.
+
+-/
+
+namespace list
+
+variables {M : Type*} [has_zero M] (l : list M)
+  [decidable_pred (λ i, nthd l i 0 ≠ 0)] (n : ℕ)
+
+/-- Indexing into a `l : list M`, as a finitely-supported function,
+where the support are all the indices within the length of the list
+that index to a non-zero value. Indices beyond the end of the list are sent to 0. -/
+def to_finsupp : ℕ →₀ M :=
+{ to_fun := λ i, nthd l i 0,
+  support := (finset.range l.length).filter (λ i, nthd l i 0 ≠ 0),
+  mem_support_to_fun := λ n, begin
+    simp only [ne.def, finset.mem_filter, finset.mem_range, and_iff_right_iff_imp],
+    contrapose!,
+    exact nthd_eq_default _ _
+  end }
```

Comment:
Is there a reason for preferring this definition over [`finsupp.on_finset`](https://leanprover-community.github.io/mathlib_docs/data/finsupp/defs.html#finsupp.on_finset)? I think the two should be defeq.

Reply:
`finsupp.on_finset` is noncomputable, while this file is all about computable constructions. Yes, they are basically the same construction, but changing the finsupp file to allow for computable constructions there is a bigger refactor. I will leave a mention in the docstring here pointing to `finsupp.on_finset`, thanks for the pointer.